File: sdk/java/src/main/java/io/omniswitch/sdk/OmniSwitchApiError.java
package io.omniswitch.sdk;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

/**
 * Every non-2xx OmniSwitch response is shaped {@code {statusCode, error,
 * code}} (validation failures add a {@code message} array instead of
 * {@code error}). {@code code} is the stable, machine-readable field
 * meant to be branched on; {@code error}/{@code message} are for
 * logging, not string-matching.
 */
public class OmniSwitchApiError extends RuntimeException {

  private final int statusCode;
  private final String code;
  private final JsonNode details;

  public OmniSwitchApiError(int statusCode, String message, String code, JsonNode details) {
    super(message);
    this.statusCode = statusCode;
    this.code = code;
    this.details = details;
  }

  public OmniSwitchApiError(int statusCode, String message, String code) {
    this(statusCode, message, code, null);
  }

  public int getStatusCode() {
    return statusCode;
  }

  public String getCode() {
    return code;
  }

  public JsonNode getDetails() {
    return details;
  }

  static OmniSwitchApiError fromResponse(int statusCode, String rawBody, ObjectMapper mapper) {
    JsonNode body = null;
    if (rawBody != null && !rawBody.isEmpty()) {
      try {
        body = mapper.readTree(rawBody);
      } catch (Exception e) {
        body = null;
      }
    }

    String message = null;
    String code = null;
    if (body != null) {
      if (body.hasNonNull("error")) {
        message = body.get("error").asText();
      } else if (body.hasNonNull("message")) {
        JsonNode messageNode = body.get("message");
        if (messageNode.isArray()) {
          StringBuilder sb = new StringBuilder();
          for (int i = 0; i < messageNode.size(); i++) {
            if (i > 0) sb.append("; ");
            sb.append(messageNode.get(i).asText());
          }
          message = sb.toString();
        } else {
          message = messageNode.asText();
        }
      }
      if (body.hasNonNull("code")) {
        code = body.get("code").asText();
      }
    }
    if (message == null) {
      message = "OmniSwitch API request failed with HTTP " + statusCode;
    }
    return new OmniSwitchApiError(statusCode, message, code, body);
  }
}

File: sdk/dotnet/src/OmniSwitchApiError.cs
using System.Text.Json;

namespace OmniSwitch.Sdk;

/// <summary>
/// Every non-2xx OmniSwitch response is shaped <c>{statusCode, error,
/// code}</c> (validation failures add a <c>message</c> array instead of
/// <c>error</c>). <see cref="Code"/> is the stable, machine-readable
/// field meant to be branched on; <c>Message</c>/the raw body are for
/// logging, not string-matching.
/// </summary>
public class OmniSwitchApiError : Exception
{
    public int StatusCode { get; }
    public string? Code { get; }
    public JsonElement? Details { get; }

    public OmniSwitchApiError(int statusCode, string message, string? code = null, JsonElement? details = null)
        : base(message)
    {
        StatusCode = statusCode;
        Code = code;
        Details = details;
    }

    internal static OmniSwitchApiError FromResponse(int statusCode, string rawBody)
    {
        JsonElement? body = null;
        if (!string.IsNullOrEmpty(rawBody))
        {
            try
            {
                body = JsonDocument.Parse(rawBody).RootElement;
            }
            catch (JsonException)
            {
                body = null;
            }
        }

        string? message = null;
        string? code = null;
        if (body is JsonElement element)
        {
            if (element.TryGetProperty("error", out var errorProp) && errorProp.ValueKind == JsonValueKind.String)
            {
                message = errorProp.GetString();
            }
            else if (element.TryGetProperty("message", out var messageProp))
            {
                if (messageProp.ValueKind == JsonValueKind.Array)
                {
                    message = string.Join("; ", messageProp.EnumerateArray().Select(e => e.GetString()));
                }
                else if (messageProp.ValueKind == JsonValueKind.String)
                {
                    message = messageProp.GetString();
                }
            }
            if (element.TryGetProperty("code", out var codeProp) && codeProp.ValueKind == JsonValueKind.String)
            {
                code = codeProp.GetString();
            }
        }
        message ??= $"OmniSwitch API request failed with HTTP {statusCode}";
        return new OmniSwitchApiError(statusCode, message, code, body);
    }
}
